### ISSUE TITLE
empty field name in array

### ISSUE THREAD

**Issue by outflux3** [n00dles/DM_Matrix#37]
I'm having a strange issue with this table; i have a wysiwyg field where i can't get it to show up, and after looking at the array, it seems that I'm getting [] without the name of the field, which is 'teaser'; 

[id] => 4
    [album_title] => Title
    [slug] => 1996-title
    [keywords] => 
    [meta_desc] => 
    [album_year] => 1996
    [album_cover] => /data/uploads/items/image.jpg
    [personnel] => 
    [] => 
    [album_description] => 
    [press_quotes] => 
    [itunes_url] => http://itunes.apple.com...

**Comment by n00dles** [comment]
can you send me your schema.xml file again. I can'r repliicate this...

**Comment by outflux3** [comment]
Hi mike,

Here's the schema.xml; I'm going to update to the latest and try resaving
all the records, later today;

-marc

**Comment by outflux3** [comment]
ok i sent it via email.
many thanks.

**Comment by outflux3** [comment]
this was the same issue as one just closed; i'll assume this is added to the tracker (something to do with the word 'teaser', or a field name that is too long, or contains an underscore)

**Comment by n00dles** [comment]
still looking into this one. can't replicate it though. 

could you zip up your whole matrix folder (if thats ok , so I can see whats wrong ) and email it to me? 
would love to figure this one out.....

**Comment by outflux3** [comment]
ok i have it - where should i email to?

**Comment by outflux3** [comment]
ok sent it...

**Comment by tablatronix** [comment]
Did you add teaser to the schema after the other fields or rather after this particular record was saved ?

**Comment by outflux3** [comment]
the original field was called album_teaser, and yes it was added later after i decided to use a separate field for the 'teaser' or summary text; and then i edited the schema in XMLSpear, to move the field up; i'm not sure if the field order matters in the data files but i moved the fields in the first data file, since i had only entered one at that point...

**Comment by tablatronix** [comment]
hmm I don't know how this works, I think the actual field names are being stored in the record files instead of a field index, that would mean that you cannot change field names after they are initially set or you will lose your data since they are not mapped to the schema but actually hardcoded per record.

**Comment by outflux3** [comment]
once i changed the field name to something without the word "teaser" it fixed the issue of an empty field name in the array, and oddly enough this happened on 2 separate completely independent sites; 
So for now i'm just making sure to stick with camelcase for field names and not use the word teaser... but it would be good to deal with the issue of re-ordering fields, which will be a persistent need during development; right now the only way to change the field order is to edit the schema file and move the field; so i'm wondering if once that is done, if you also need to re-order the fields within the data record, or if that doesn't matter and the data in each record is more of a bucket concept..

**Comment by tablatronix** [comment]
I checked in an underscore fix yesterday, give it a whirl.

I cant imagine how the word teaser would be the issue, it is probably more related t the changing of field names or your reordering of indexes.

**Comment by outflux3** [comment]
ok cool; this particular site has been changed to all camelcase now so works, but still has the double id field problem, as well as some strangeness with CDATA inconsistencies in the data files.. 

i will update matrix now on the current project, because i'm still using underscores for field names there, and that's where i wasn't able to use the WHERE statement (used an array filter as a stopgap fix..); 

besides the schema and the data files, is there some other place that things are being stored, i.e. cache, that i can clear?

**Comment by tablatronix** [comment]
I have the same question for noodles, since my old tables are missing drop table functionality and stuff, so working with old datafiles with alpha mutating code is naturally going to have issues. And i am unsure how he engineered the schema and record tables.

**Comment by n00dles** [comment]
The only thing that might be messed up over the last few changes was the duplicate ID field which will be fixed once I update the latest revision. 

Other than that the only other change that will be made tonight will be to the main schema.xml file and 2 new attributes added to each field 'size' and 'visibility' 

Size will be use to allow fluid type layouts for most field types. The visibility option will allow you to auto open/close fields by default or condition of a value in the field.

**Comment by n00dles** [comment]
Drop table option is only visible for tables with no records to prevent accidental deletion of the table and records by the end users. I'll add a drop/delete function some time this evening.

**Comment by n00dles** [comment]
As for changing field names,  your right if you change a field name when records exists it will screw things up. 

I'll add code similar to the change Url code in GS to fix this, and loop through each of the records and change the name. 

reordering fields in the record file doesn't really matter once the correspond to the field names in the main schema.xml file.

**Comment by tablatronix** [comment]
wow loop through every record file to rename a field. too bad they arent just linked index ids.

**Comment by tablatronix** [comment]
I have the gray scale versions of the silk icons, maybe we can throw disabled icons in there, so people know they are not broken.

**Comment by n00dles** [comment]
I don't see this as a huge issue as its not going to be done very often. If you've a table with a 1000 records that's working fine and you've built your site around it, why would you rename ? 

I'll look at the indexing. see if its doable. Don't want to spend a huge amount of time on it as I don't see it being used that much.

**Comment by n00dles** [comment]
@tablatronix, yeah that sound good with the icons.

**Comment by tablatronix** [comment]
no I woudln't spend the time, it would require too much work for initial release. besides you can always add a maintenance job to clean up some stuff and resync records if someone needed to rework their schemas.

**Comment by n00dles** [comment]
What do you guys thing of a lockdown password which would disable the editing/renaming/deting of tables from the end user? 

That was we can deign the table and just allows adding/delting of records. last thing I want is end users delting important table or screwing things up.

**Comment by outflux3** [comment]
maybe we just prevent table deletion entirely, because if you have to really delete a table, you could delete the folder and then directly edit the schema.xml file to remove that table's entry;

so it would be something the site developer would need to do;

as far as editing tables and fields, is there a way to allow for re-ordering of the fields, which i can see being needed a lot, say you have a table and then the client needs a new field but it needs to be in a specific place in the order of fields; and i can also see the occasional need to rename or delete fields; so maybe there just needs to be some guidelines for how to do that in the XML files, so a developer could work with an application and not get stuck with some decision made early on..

i used MassReplaceIt and was able to rename all of the fields from having underscores to being camelcase, in a few minutes;

**Comment by n00dles** [comment]
I have reordering about 50% done. I'll finish it off after the visibility additon which shouldn't take too long.

**Comment by outflux3** [comment]
visibility will be great...kind of like in Processwire where you can choose how the field shows on the editor screen...

when i was changing field names, i was always doing in directly in the xml files, and then changing both the schema and the records, using mass replaceit

**Comment by n00dles** [comment]
exactly like PW 8) 
That's were the inspiration and UI came from

**Comment by outflux3** [comment]
yeah, processwire is great, but sometimes a bit much for really small simple sites;  i'll use it for when the client needs really complex custom content, repeating fields.... i also love that image upload drag and drop thing right on the editor screen. GS could really use some better file management, since that's a major complaint now, in that the images have to be uploaded first from the files tab, and then selected within the wysiwyg pop up..

**Comment by tablatronix** [comment]
File management in GS is something I plan on working on soon , provided i can get discussions going with the rest of the developers right now its not very active.
I have already implemented in ckeditor file uploads on me test builds, but with the lack of proper branches etc, its a pain to work on these things, hopefully things improve when we get over to github. v3.2 should have in editor file uploads and maybe drag and drop capability.

**Comment by n00dles** [comment]
Yeah we've been a little slow lately with our real jobs and this stuff to be able to spend much time on the core. 
Much preferring Github than Google code, and would love to move it over here. I'll bring it up with Chris again and see what plans are....

**Comment by outflux3** [comment]
@tablatronix - sounds great...; editor file uploads will make a tremendous difference for the client workflow;
github is great, this is the first time i have used it (mostly as observer/tester), but seems more intuitive than GC

**Comment by n00dles** [comment]
If you get a chance please test the latest updates. you will need to save 1 record I think to add the new options to the schema.xml. other than that there should be no changes. 

I'll add a schema version from now to take care of any other changes that might occur and we can update automatically.

**Comment by outflux3** [comment]
i updated both sites - seems like the field list lost the alternate row color?
i tried saving the table and some of the records but haven't checked the xml files..can do that later tonight (ny time)

**Comment by n00dles** [comment]
cool thanks, I think the table Drag and drop is overriding it.

**Comment by n00dles** [comment]
just updated the code with field reordering. please test if ye get a chance. 
I'll add some proper styling to the draggable rows when I get a chance.

**Comment by outflux3** [comment]
wow - loving the new options, the reorder is awesome..
the field width is also great, makes the data entry really nice!
all looking really slick...many thanks.

**Comment by n00dles** [comment]
Ok I've removed field rename if the table has records, for the moment until I get some time to code in the rename field function.

**Comment by lokothodida** [comment]
Leaving this open for now in case there are problems, but since the overhaul I've coded it so that when fields are renamed, it changes the field names in the database too (as n00dles had planned to do).

Also, in the event of any troubles, remember that there is now a Backup and a Restore option for your tables, so that you can roll them back to a prior state if a backup exists. Use it when you're experimenting with your tables.
